### PR TITLE
Sift algorithm

### LINKED ISSUE #967: Implementing SIFT for detect local features of image.
# Description

- In this script, we are going to implement SIFT(Scale Invariant Feature Transform) algorithm of `Open CV`.
- The SIFT features are local and based on the appearance of the object at particular interest points, and are invariant to image scale and rotation.

## Language

- [x] Python

## Checklist:

- **Details on the Script**

| name            | about                                                  | labels  | assignees                                                      |
| --------------- | ------------------------------------------------------ | ------- | -------------------------------------------------------------- |
| SIFT algorithm | We use Python to implement SIFT algorithm of Open CV to detect the feature of image | GSSOC21 Python | @ShubhamGupta577 |

### PR DESCRIPTION
**Pull request by ShubhamGupta577**
# Description

- In this script, we implemented SIFT(Scale Invariant Feature Transform) algorithm of Open CV using Python.
- This algorithm is used to detect the various feature of an image.

Fixes #967 

## Type of change
- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines(Clean Code) of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have created a helpful and easy to understand `README.md`
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests/screenshots(if any) that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by ShubhamGupta577** [comment]
@HarshCasper, Please review this PR.

File: SIFT Algorithm/SIFT_Algorithm.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+kp, dp = sift.detectAndCompute(gray, None)
+
+img=cv2.drawKeypoints(gray,kp,flags = cv2.DRAW_MATCHES_FLAGS_DRAW_RICH_KEYPOINTS)
+
+# Print the number of keypoints detected in the image
+print("Number of Keypoints Detected In The Image: ", len(kp))
+
+#Display the final output image
+cv2_imshow('sift_keypoints.jpg',img)
```
Comment (HarshCasper) [inline SIFT Algorithm/SIFT_Algorithm.py:34]: Can we have a dynamic output image name as well?

File: SIFT Algorithm/SIFT_Algorithm.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+#Detect keypoints and Create Descriptor
+sift = cv2.xfeatures2d.SIFT_create()
+
+kp, dp = sift.detectAndCompute(gray, None)
+
+img=cv2.drawKeypoints(gray,kp,flags = cv2.DRAW_MATCHES_FLAGS_DRAW_RICH_KEYPOINTS)
+
+# Print the number of keypoints detected in the image
+print("Number of Keypoints Detected In The Image: ", len(kp))
```
Comment (HarshCasper) [inline SIFT Algorithm/SIFT_Algorithm.py:31]: Put all of the code in a modular function and call the function from a driver code

**Review by HarshCasper — CHANGES_REQUESTED** [review]

**Comment by sohamsshah** [comment]
@ShubhamGupta577 IMO pure and direct algorithm scripts are less preferable. If you can create a utilitarian script by making use of this algorithm, I suppose it can provide more value to the project. You can search for applications of this algorithm and then implement. Idea is to have scripts that are helpful in some way to the developers/users. 

I hope you understand, thanks.

**Review by vybhav72954 — CHANGES_REQUESTED** [review]

**Comment by ShubhamGupta577** [comment]
@sohamsshah, I understand your concern but IMO the script which I added is more clear and helpful for the users/developers because whoever needs this script is directly available here otherwise they have to extract the useful code from this script which will be more tedious.

File: SIFT Algorithm/Readme.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+- The SIFT features are local and based on the appearance of the object at particular interest points, and are invariant to image scale and rotation.
+
+## Setup instructions
+
+- This algorithm is patented, so this algorithm is included in the Non-free module in OpenCV.
+- For using this script, you need to use the `Open CV version <3.0` and additonally you need to install the `opencv-contrib-python` of the same version.
+
+## Output
+
+<img src="https://i.ibb.co/SJ1rfC3/Input-Image.png" width=450/> <img src="https://i.ibb.co/VHgMqWN/Output-image.png" width=450/>
```
[the remark this answers is not in the document]
Comment (ShubhamGupta577) [reply SIFT Algorithm/Readme.md:13]: I had to resize the image for the readme and by using markdown I was not able to achieve it. So I had used HTML tags for the same purpose.

**Comment by ShubhamGupta577** [comment]
@HarshCasper, Requested changes had done. Please review the script.

**Comment by sohamsshah** [comment]
> @sohamsshah, I understand your concern but IMO the script which I added is more clear and helpful for the users/developers because whoever needs this script is directly available here otherwise they have to extract the useful code from this script which will be more tedious.

I appreciate your efforts for working on this. But, this script is pure algorithm implementation. SIFT is a really good algorithm which is used at various CV applications, but pure algorithm implementations aren't applicable to the niche of Rotten Scripts.

As I said, if you can bring some use-case script that uses one or more CV concepts to solve a problem, it is readily welcomed. I hope you understand.

**Comment by ShubhamGupta577** [comment]
@sohamsshah, I understand what you say, As SIFT is invariance to image scale and rotation for the feature detection of an image. So as an addition to the script can we add the code to
- Generate a test image by adding Scale invariance and rotational invariance to the input image which changes the alignment and resolution of the image.
- Then apply  **Brute Force Matching Algorithm of Open CV** for matching the features of input and test image. This will show the application of SIFT.

If it makes our script useful, please let me know.

**Comment by sohamsshah** [comment]
> @sohamsshah, I understand what you said, As SIFT is invariance to image scale and rotation for the feature detection of an image. So as an addition to the script can we add the code to
> 
> * Generate a test image by adding Scale invariance and rotational invariance to the input image which changes the alignment and resolution of the image.
> * Then apply  **Brute Force Matching Algorithm of Open CV** for matching the features of input and test image. This will show the application of SIFT.
> 
> If it makes our script useful, please let me know.

Okay, sounds good. You can go ahead and rename the script accordingly. Also, try to produce enough examples and descriptive README for the same for the users to understand the script's utility clearly.

Thanks for working on this

**Comment by ShubhamGupta577** [comment]
@sohamsshah, I had added the code of Brute Force Matcher for matching the features of the input and test image and also updated the readme file.
Please review this PR.

**Review by sohamsshah — CHANGES_REQUESTED** [review]
Good work @ShubhamGupta577 !

Few small suggestions from my side:
- I see that the `SIFT ALGORITHM` file is put in the root folder of the repo. Can you shift this directory under Python? 
- I would suggest to add 2-3 more `input` `output` examples to the Readme.
- You can also share the CV pipeline (a brief flow of how various CV operations lead to the final output) in the README.md

After that, its good to merge. 
Again thanks for working on this :)

**Review by sohamsshah — APPROVED** [review]
Great work @ShubhamGupta577 
LGTM @HarshCasper

File: Python/SIFT Algorithm/SIFT_Algorithm.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
+'''
+    SIFT(Scale Invariant Feature Transform) Algorithm for the feature detection.
+    This algorithm is patented, so this algorithm is included in the Non-free module in OpenCV.
+'''
+
+#Importing required libraries
+import cv2
+import numpy as np
+
+def siftAlgorithm():
```
Comment (vybhav72954) [inline Python/SIFT Algorithm/SIFT_Algorithm.py:10]: Missing a docstring here

File: Python/SIFT Algorithm/SIFT_Algorithm.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+    result = cv2.drawMatches(image, kp, test_image, test_kp, matches, test_gray, flags = 2)
+
+    # Display the best matching points
+    cv2.imshow('Feature matching',result)
+
+    # Print total number of matching points between the training and query images
+    print("\nNumber of Matching Keypoints Between The Training and Query Images: ", len(matches))
+
+#Driver Code
```
Comment (vybhav72954) [inline Python/SIFT Algorithm/SIFT_Algorithm.py:71]: Can we trigger this using the proper PIP guide?

Something like
```
if __name__ == "__main__"
```

**Review by vybhav72954 — CHANGES_REQUESTED** [review]

**Review by vybhav72954 — APPROVED** [review]

**Review by HarshCasper — APPROVED** [review]

**Pull request merged**